clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorRequest.java

```diff
@@ -0,0 +1,145 @@
+            bld.append("(type=FindCoordinatorRequest, coordinatorKey=");
+            bld.append(coordinatorKey);
+            bld.append(", coordinatorType=");
+            bld.append(coordinatorType);
+            bld.append(")");
+            return bld.toString();
+        }
+    }
+
+    private final String coordinatorKey;
+    private final CoordinatorType coordinatorType;
+
+    private FindCoordinatorRequest(CoordinatorType coordinatorType, String coordinatorKey, short version) {
+        super(version);
+        this.coordinatorType = coordinatorType;
+        this.coordinatorKey = coordinatorKey;
+    }
+
+    public FindCoordinatorRequest(Struct struct, short version) {
+        super(version);
+
+        if (version == 0) {
+            this.coordinatorType = CoordinatorType.GROUP;
+            this.coordinatorKey = struct.getString(GROUP_ID_KEY_NAME);
+        } else {
+            this.coordinatorType = CoordinatorType.forId(struct.getByte(COORDINATOR_TYPE_KEY_NAME));
+            this.coordinatorKey = struct.getString(COORDINATOR_KEY_KEY_NAME);
+        }
+    }
+
+    @Override
+    public AbstractResponse getErrorResponse(Throwable e) {
+        short versionId = version();
+        switch (versionId) {
+            case 0:
+            case 1:
+                return new FindCoordinatorResponse(Errors.forException(e), Node.noNode());
+
+            default:
+                throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
+                        versionId, this.getClass().getSimpleName(), ApiKeys.FIND_COORDINATOR.latestVersion()));
+        }
+    }
+
+    public String coordinatorKey() {
+        return coordinatorKey;
+    }
+
+    public CoordinatorType coordinatorType() {
+        return coordinatorType;
+    }
+
+    public static FindCoordinatorRequest parse(ByteBuffer buffer, short version) {
+        return new FindCoordinatorRequest(ApiKeys.FIND_COORDINATOR.parseRequest(version, buffer), version);
+    }
+
+    @Override
+    protected Struct toStruct() {
+        short version = version();
+        Struct struct = new Struct(ApiKeys.FIND_COORDINATOR.requestSchema(version));
+        if (version == 0) {
+            struct.set(GROUP_ID_KEY_NAME, coordinatorKey);
+        } else {
+            struct.set(COORDINATOR_TYPE_KEY_NAME, coordinatorType.id);
+            struct.set(COORDINATOR_KEY_KEY_NAME, coordinatorKey);
+        }
+        return struct;
+    }
+
+    public enum CoordinatorType {
+        GROUP((byte) 0), TRANSACTION((byte) 1), UNKNOWN((byte) -1);
+
+        final byte id;
+
+        CoordinatorType(byte id) {
+            this.id = id;
+        }
+
+        public static CoordinatorType forId(byte id) {
+            switch (id) {
+                case 0:
+                    return GROUP;
+                case 1:
+                    return TRANSACTION;
+                default:
+                    return UNKNOWN;
```

Comment:
I generally prefer to throw in such cases. We should probably do the same for `TransactionResult`. I wanted to raise that in the other PR, but then forgot.